compiler/rustc_codegen_llvm/src/consts.rs

```diff
@@ -171,8 +172,28 @@ fn check_and_apply_linkage<'ll, 'tcx>(
     if let Some(linkage) = attrs.import_linkage {
         debug!("get_static: sym={} linkage={:?}", sym, linkage);
 
-        // Declare a symbol `foo` with the desired linkage.
-        let g1 = cx.declare_global(sym, cx.type_i8());
+        // Declare a symbol `foo`. If `foo` is an extern_weak symbol, we declare
+        // an extern_weak function, otherwise a global with the desired linkage.
+        let g1 = if matches!(attrs.import_linkage, Some(Linkage::ExternalWeak)) {
+            // An `extern_weak` function is represented as an `Option<unsafe extern ...>`,
+            // we extract the function signature and declare it as an extern_weak function
+            // instead of an extern_weak i8.
+            let instance = Instance::mono(cx.tcx, def_id);
+            if let ty::Adt(struct_def, generic_args) = instance.ty(cx.tcx, cx.typing_env()).kind()
+                && matches!(cx.tcx.get_diagnostic_name(struct_def.did()), Some(sym::Option))
```

Comment:
`get_diagnostic_name` must not be used for anything other than diagnostics. You should use a lang item here instead. Instead you should check if it the `DefId` matches the `DefId` of the `Option` lang item (make sure to also handle the case where the lang item doesn't exist)

Reply:
Oh thanks for the pointer, that makes sense!

Reply:
there's is_lang_item too

Reply:
Ahh, that's even nicer, thanks!